apollo-federation/src/query_plan/query_planner.rs

```diff
@@ -490,11 +500,179 @@ impl QueryPlanner {
-    _parameters: &QueryPlanningParameters,
-    _has_defers: bool,
+    parameters: &QueryPlanningParameters,
+    has_defers: bool,
 ) -> Result<Vec<FetchDependencyGraph>, FederationError> {
-    todo!("FED-127")
+    // FINDME
+    let QueryPlanningParameters {
+        supergraph_schema,
+        federated_query_graph,
+        operation,
+        ..
+    } = parameters;
+    let root_type: Option<CompositeTypeDefinitionPosition> = if has_defers {
+        supergraph_schema
+            .schema()
+            .root_operation(operation.root_kind.into())
+            .and_then(|name| supergraph_schema.get_type(name.clone()).ok())
+            .and_then(|ty| ty.try_into().ok())
+    } else {
+        None
+    };
+    // We have to serially compute a plan for each top-level selection.
+    let mut split_roots = split_top_level_fields(operation.selection_set.clone());
+    let mut digest = Vec::new();
+    let mut starting_fetch_id = 0;
+    let selection_set = split_roots
+        .next()
+        .ok_or_else(|| FederationError::internal("Empty top level fields"))?;
+    let BestQueryPlanInfo {
+        mut fetch_dependency_graph,
+        path_tree: mut prev_path,
+        ..
+    } = compute_root_parallel_best_plan(parameters, selection_set, has_defers)?;
+    let mut prev_subgraph = only_root_subgraph(&fetch_dependency_graph)?;
+    for selection_set in split_roots {
+        let BestQueryPlanInfo {
+            fetch_dependency_graph: new_dep_graph,
+            path_tree: new_path,
+            ..
+        } = compute_root_parallel_best_plan(parameters, selection_set, has_defers)?;
+        let new_subgraph = only_root_subgraph(&new_dep_graph)?;
+        if new_subgraph == prev_subgraph {
+            // The new operation (think 'mutation' operation) is on the same subgraph than the previous one, so we can concat them in a single fetch
+            // and rely on the subgraph to enforce seriability. Do note that we need to `concat()` and not `merge()` because if we have
+            // mutation Mut {
+            //    mut1 {...}
+            //    mut2 {...}
+            //    mut1 {...}
+            // }
+            // then we should _not_ merge the 2 `mut1` fields (contrarily to what happens on queried fields).
+
+            prev_path = OpPathTree::merge(&prev_path, &new_path);
+            fetch_dependency_graph = FetchDependencyGraph::new(
+                supergraph_schema.clone(),
+                federated_query_graph.clone(),
+                root_type.clone(),
+                starting_fetch_id,
+            );
+            compute_root_fetch_groups(
+                operation.root_kind,
+                &mut fetch_dependency_graph,
+                &prev_path,
+            )?;
+        } else {
+            // TODO(@TylerBloom): Verify that this is correct? It could be the case that we just
+            // need to grab `.starting_id_generation`.
+            starting_fetch_id = fetch_dependency_graph.next_fetch_id();
```

Comment:
I was unsure if JS code generated the next fetch ID, used the starting ID

Reply:
yeah confusingly JS code does not increment the ID in `next_fetch_id()` (unless I'm missing something it looks like the only place that actually increments the ID is in `extractChildrenAndDeferredDependencies`) but I don't think it matters much? instead of starting with 0 we would just start with 1?

Reply:
There is a chance that `start_fetch_id` gets called multiple times. Also, because we merge dep fetch graph, I'm not sure if their starting number changes (I don't think so, but I'm not 100% sure). In which case, we might want to start at, say, 10 but instead we start at 12 (should this path be taken more than once).

Reply:
I am going to stick with this approach. I'd left a `PORT_NOTE` comment for reference should this cause issues in the future.